src/fieldExtensions.js

```diff
@@ -0,0 +1,200 @@
+/**
+ * The flattenState() field extension allows the user to access fields normally like
+ * `state.field` instead of `state.present.field`.
+ *
+ * Warning: if your state has fields that conflict with redux-undo's like .past or .index
+ * they will be overridden. You must access them as `state.present.past` or `state.present.index`
+ */
+
+export const flattenState = () => {
+  return (undoableConfig) => {
+    console.log(undoableConfig)
+    if (!undoableConfig.disableWarnings) {
+      console.warn(
+        'Warning: the flattenState() extension prioritizes redux-undo fields when flattening state.',
+        'If your state has the fields `limit` and `present`, you must access them',
+        'with `state.present.limit` and `state.present.present` respectively.\n',
+        'Only works with objects as state. Do not use flattenState() with primitives or arrays.\n',
+        'Disable this warning by passing `disableWarnings: true` into the undoable config'
+      )
+    }
+
+    return (state) => {
+      // state.present MUST be spread first so that redux-undo fields have priority
+      return { ...state.present, ...state }
+    }
+  }
+}
+
+/**
+ * @callback actionFieldIncludeAction
+ * @param {Action} action - The current action.
+ * @returns {boolean}
+ */
+
+/**
+ * The actionField() field extension allows users to insert the last occuring action
+ * into their state.
+ *
+ * @param {Object} config - Configure actionField()
+ *
+ * @param {string} config.insertMethod - How the last action will be inserted. Possible options are:
+ *   - actionType: { ...state, actionType: 'LAST_ACTION' }
+ *   - action: { ...state, action: { type: 'LAST_ACTION', ...actionPayload } }
+ *   - inline: { ...state, present: { action: { type: 'LAST', ...payload }, ...otherFields } }
+ *
+ * @param {actionFieldIncludeAction} config.includeAction - A filter function that decides if
+ * the action is inserted into history.
+ */
+export const actionField = ({ insertMethod, includeAction } = {}) => {
+  if (insertMethod === 'inline') {
+    return inlineActionField({ includeAction })
+  }
+
+  let extend
+  if (insertMethod === 'action') {
+    extend = (state, action) => ({ ...state, action })
+  } else if (!insertMethod || insertMethod === 'actionType') {
+    extend = (state, action) => ({ ...state, actionType: action.type })
+  } else {
+    throw new Error(
+      `Unrecognized \`insertMethod\` option for actionField() extension: ${insertMethod}.\n` +
+      'Options are "action", "inline", or "actionType"'
+    )
+  }
+
+  return (undoableConfig) => {
+    const included = includeAction || (() => true)
+
+    if (!undoableConfig.disableWarnings) {
+      console.warn(
+        'Warning: the actionField() extension might override other state fields',
```

Comment:
also here, should we only warn when we detect that it is overriding something?

Reply:
Same idea as above.